### PR TITLE
Change Trie::save() method

### PR DESCRIPTION
**Pull request by julianlen**

### CODE REVIEW

**Comment by tinchou** [comment]
Consider changing `NodeReference#save` to:

```java
public void save() {
    if (lazyNode != null && !isEmbeddable()) {
        lazyNode.save();
    }
}
```

File: rskj-core/src/main/java/co/rsk/trie/NodeReference.java [hunk 1]
```diff
@@ -55,7 +54,11 @@ public boolean isEmpty() {
      */
     public void save() {
         if (lazyNode != null) {
-            lazyNode.save();
+            if (!isEmbeddable()) {
+                lazyNode.save();
+            } else {
+                store.saveValue(lazyNode);
```
Comment (collheesi) [inline rskj-core/src/main/java/co/rsk/trie/NodeReference.java]: This is called even when the node doesn't have a long value. I think this is a bug.

File: rskj-core/src/main/java/co/rsk/trie/Trie.java [hunk 2]
```diff
@@ -503,6 +507,10 @@ public Trie delete(String key) {
      * @return a byte array with the serialized info
      */
     public byte[] toMessage() {
+        if (encoded != null) {
+            return cloneArray(encoded);
```
Comment (SergioDemianLerner) [inline rskj-core/src/main/java/co/rsk/trie/Trie.java]: Doesn't cloning affect performance ? Trie.gets() also clone. Can't we just tag immutable arrays with large banners in comments and trust humans to follow the specified semantic ?

File: rskj-core/src/main/java/co/rsk/trie/Trie.java [hunk 3]
```diff
@@ -563,7 +571,13 @@ public Trie delete(String key) {
...
         }
 
-        return buffer.array();
+        encoded = buffer.array();
+
+        return cloneArray(encoded);
+    }
+
+    public int getMessageLength() {
+        return toMessage().length;
```
Comment (SergioDemianLerner) [inline rskj-core/src/main/java/co/rsk/trie/Trie.java]: unnecessary clone here.

File: rskj-core/src/main/java/co/rsk/trie/Trie.java (hunk 2 above)
Comment (collheesi) [reply rskj-core/src/main/java/co/rsk/trie/Trie.java]: I don't think we can trust humans. We plan to take the Pantheon approach and have immutable objects that can be safely passed around (`BytesValue`), but we can't do it until now without interfering with other humans' schedules.

See: https://github.com/PegaSysEng/pantheon/blob/master/util/src/main/java/tech/pegasys/pantheon/util/bytes/BytesValue.java

File: rskj-core/src/main/java/co/rsk/trie/Trie.java (hunk 3 above)
Comment (lsebrie) [reply rskj-core/src/main/java/co/rsk/trie/Trie.java]: We made this change to address this clone https://github.com/rsksmart/rskj/pull/874

**Pull request merged**